### ISSUE TITLE
Sync version in custom package.json properties

### ISSUE THREAD

**Issue by LudovicSterlin** [JamieMason/syncpack#112]
## Description

Sync version in custom `package.json` properties to match the version in `(dev/peer/optional)dependencies`.
The custom properties can be nested propertyies of package.json.

For example, [directus](https://directus.io/) has a utility that creates a `package.json` with a `directus:extension` property like the below:
```json
{
  [...]
  "directus:extension": {
    "type": "endpoint",
    "path": "index.js",
    "source": "src/index.ts",
    "host": "9.22.2"
  },
  [...]
```
Where the value at `directus:extension.host` is actually a [directus](https://www.npmjs.com/package/directus) version

## Suggested Solution

Add a `dependenciesCustomPaths` option in the form of an array of custom dependencies paths each defined by the name of the dependencies it should be associated with and its path in the `package.json` file.

In TypeScript this would result in something like:
```ts
dependenciesCustomPaths: { name: string; path: string }[];
```

This option could be given through the config file, if we take the example from above this would result in a `.syncpackrc.yaml` like below:
```yaml
dependenciesCustomPaths:
  - name: 'directus'
    path: 'directus:extension.host'
```

## Help Needed

A bit of help on testing other syncpack features than `list/fix-mismatches` which are the ones I am familiar with.

Review of the associated MR -> #113

**Comment by JamieMason** [comment]
Thanks for this @LudovicSterlin, I've had an idea and am just leaving some notes here while I still have it.

The internals could be refactored to use this data structure internally, which can't be changed but can be added to:

```ts
const dependencyTypesByName = {
  dependencies: {
    strategy: 'versionsByName',
    path: 'dependencies',
  },
  devDependencies: {
    strategy: 'versionsByName',
    path: 'devDependencies',
  },
  overrides: {
    strategy: 'versionsByName',
    path: 'overrides',
  },
  peerDependencies: {
    strategy: 'versionsByName',
    path: 'peerDependencies',
  },
  pnpmOverrides: {
    strategy: 'versionsByName',
    path: 'pnpm.overrides',
  },
  resolutions: {
    strategy: 'versionsByName',
    path: 'resolutions',
  },
  workspace: {
    strategy: 'name~version',
    namePath: 'name',
    versionPath: 'version',
  },
};
```

Users could add their own in `.syncpackrc` which would be merged with the internal ones above to direct syncpack where to find and fix other versions.

```jsonc
{
  "dependencyTypesByName": {
    "packageManager": {
      "strategy": "versionsByName",
      "path": "packageManager"
    },
    "engines": {
      "strategy": "versionsByName",
      "path": "engines"
    },
    // this is a more granular alternative to the above
    "enginesNodeOnly": {
      "strategy": "version",
      "path": "engines.node"
    },
    "directusHost": {
      "strategy": "version",
      "path": "directus:extension.host"
    }
  }
}
```

Possible types of `strategy`:

- `versionsByName` A typical list of dependencies
  ```json
  {
    "dependencies": {
      "foo": "1.2.3"
    }
  }
  ```
- `name~version` A name and version which are siblings, like local packages:
  ```json
  {
    "name": "foo",
    "version": "1.2.3"
  }
  ```
- `name@version` A name/version pair found in a single value
  ```json
  {
    "packageManager": "pnpm@1.2.3"
  }
  ```
- `version` An anonymous version found in a single value
  ```jsonc
  {
    "directus:extension": {
      "host": "9.22.2"
    },
    // OR
    "engines": {
      "node": "1.2.3"
    }
  }
  ```

This data could be referenced in version and semver groups etc.

**Comment by JamieMason** [comment]
This is now available as of [9.0.0](https://github.com/JamieMason/syncpack/releases/tag/9.0.0), please let me know if you run into any problems – thanks a lot.
